docs/adapters/adapter_api.md

```diff
@@ -6,9 +6,64 @@ NOTE: This API is currently under active development and is highly likely to cha
+
+For reads, the adapter's job is to interpret the options included in the juttle `read` command into a set of matching data points, construct Javascript objects representing those data points, and pass them to the juttle runtime. For writes, the adapter's job is to take the output of programs and "save" that output by writing data points to the backend.
+
+More sophisticated adapters work together with the juttle optimizer to push aggregation operations directly into the backend. For example, to count the number of data points in a given time period you could simply fetch all the data points and have the juttle program perform the counting. However, if supported by the backend it would be more efficient to count the number of data points directly in the backend and return the count instead.
+
+## Writing an Adapter
+
+Here are the main steps involved in writing a new adapter:
+
+1. Create a javascript module that can be require()d by the juttle runtime.
+2. From that module, export an initialization function to define the read/write capabilities of the adapter.
+3. Implement ES6 classes deriving from AdapterRead/AdapterWrite that perform the work of reading from/writing to the backend data source.
+
+We'll go through each of these steps in more detail below.
+
+## Javascript Modules, Classes and Methods
+
+We've created the skeleton of an adapter in the [juttle-adapter-template](https://github.com/juttle/juttle-adapter-template) repository on github. That repository is a good reference for the layout of files, classes, functions, etc.
+
+An adapter is packaged as a npm module. When the adapter is loaded, the CLI/juttle-engine perform a `require` of the module (i.e. the top-level directory). In [package.json](https://github.com/juttle/juttle-adapter-template/blob/master/package.json), the module defines a `main` of [lib/index.js](https://github.com/juttle/juttle-adapter-template/blob/master/lib/index.js), so `lib/index.js` is loaded.
+
+The main function exported by `lib/index.js` takes a `config` argument containing the configuration object for the adapter, and returns an object with `name`, `read`, `write`, and `optimize` attributes:
+
+```
+{
+    name: <adapter name>
+    read: <read adapter class>,
+    write: <write adapter class>,
+    optimize: <optimization module>
+}
+```
+
+The value for the `name` attribute corresponds to the `read <name>`/`write <name>` in juttle programs. The value for `read` is a class inheriting from [AdapterRead](../../lib/adapters/adapter-read.js) which performs the read work of the adapter. The value for `write` is an class inheriting from [AdapterWrite](../../lib/adapters/adapter-write.js) which performs the write work of the adapter. The value for `optimizer` is a module exporting functions `optimize_head`, `optimize_tail`, and `optimize_reduce` that do the work of optimizing the program.
```

Comment:
I took all the ones listed in the template. I don't see optimize_sort in lib/compiler/optimize. Is there somewhere else I can find the authoritative set?